File: backend/src/AURA.Api/Contracts/ApiEnvelope.cs
namespace AURA.Api.Contracts;

public sealed record ApiError(string Code, string Message, object? Details = null);

public sealed record ApiEnvelope<T>(bool Success, T? Data, ApiError? Error, string TraceId)
{
    public static ApiEnvelope<T> Ok(T data, string traceId) => new(true, data, null, traceId);
    public static ApiEnvelope<T> Fail(ApiError error, string traceId) => new(false, default, error, traceId);
}

File: backend/src/main/java/com/aura/backend/common/response/ApiEnvelope.java
package com.aura.backend.common.response;

import java.util.UUID;

public record ApiEnvelope<T>(boolean success, T data, ApiError error, String traceId) {

    public static <T> ApiEnvelope<T> success(T data) {
        return new ApiEnvelope<>(true, data, null, UUID.randomUUID().toString());
    }

    public static ApiEnvelope<Void> failure(ApiError error) {
        return new ApiEnvelope<>(false, null, error, UUID.randomUUID().toString());
    }
}
